File: pages/api/analyze.ts
import type { NextApiRequest, NextApiResponse } from 'next';

const FINNHUB_API_KEY = 'd1epc69r01qghj42aj4gd1epc69r01qghj42aj50';
const FINNHUB_BASE = 'https://finnhub.io/api/v1';

type Option = {
  symbol: string;
  strike: number;
  lastPrice: number;
  impliedVolatility: number;
  openInterest: number;
  volume: number;
  type: string;
};

type ChartPoint = { price: number; profit: number };

type ApiResponse = {
  summary: string;
  top_5: string[][];
  chartData: ChartPoint[];
  error?: string;
};

async function fetchOptions(symbol: string): Promise<{ expiry: string; options: Option[] } | null> {
  const expRes = await fetch(`${FINNHUB_BASE}/stock/option-chain?symbol=${symbol}&token=${FINNHUB_API_KEY}`);
  if (!expRes.ok) return null;
  const expData = await expRes.json();
  if (!expData.data || !expData.data.length) return null;
  // Use the first expiration date for demo
  const expiryObj = expData.data[0];
  const expiry = expiryObj.expirationDate;
  const options = Array.isArray(expiryObj.options) ? expiryObj.options : [];
  if (!options.length) return null;
  return { expiry, options: options as Option[] };
}

export default async function handler(req: NextApiRequest, res: NextApiResponse<ApiResponse | { error: string }>) {
  if (req.method !== 'POST') {
    res.status(405).json({ error: 'Only POST allowed' });
    return;
  }
  try {
    const { ticker, strategy } = req.body;
    if (!ticker || !['bullish', 'bearish'].includes(strategy)) {
      res.status(400).json({ error: 'Invalid input' });
      return;
    }
    const symbol = ticker.toUpperCase();
    const result = await fetchOptions(symbol);
    if (!result) {
      res.status(200).json({ summary: '', top_5: [], chartData: [], error: 'No options data found for this symbol.' });
      return;
    }
    const { expiry, options } = result;
    let filtered = options.filter((opt) => opt.type === (strategy === 'bullish' ? 'CALL' : 'PUT'));
    filtered = filtered.filter((opt) => opt.impliedVolatility && opt.strike && opt.lastPrice);
    filtered.sort((a, b) => b.impliedVolatility - a.impliedVolatility);
    const top5 = filtered.slice(0, 5);
    if (top5.length === 0) {
      res.status(200).json({ summary: '', top_5: [], chartData: [], error: 'No suitable options found for this strategy.' });
      return;
    }
    const top_5 = top5.map((opt) => [
      opt.symbol,
      String(opt.strike),
      String(opt.lastPrice),
      String(opt.impliedVolatility),
      String(opt.openInterest),
      String(opt.volume),
      expiry
    ]);
    const strikes = top5.map((opt) => Number(opt.strike));
    const chartData: ChartPoint[] = [];
    for (let i = 0; i < 20; ++i) {
      const price = Math.round((Math.min(...strikes) * 0.9 + (Math.max(...strikes) * 1.1 - Math.min(...strikes) * 0.9) * (i / 19)) * 100) / 100;
      const profit = strategy === 'bullish'
        ? (price - strikes[0]) * 10
        : (strikes[0] - price) * 10;
      chartData.push({ price, profit: Math.round(profit * 100) / 100 });
    }
    const summary = `Top ${strategy.charAt(0).toUpperCase() + strategy.slice(1)} trade for ${symbol}: ${top_5[0][0]} (Strike ${top_5[0][1]}, IV ${top_5[0][3]})`;
    res.status(200).json({ summary, top_5, chartData });
  } catch (e) {
    res.status(500).json({ error: e instanceof Error ? e.message : 'Internal error' });
  }
} 

File: api/analyze.py
import json
import time
import yfinance as yf
import pandas as pd
import numpy as np
from scipy.stats import norm

def handler(request):
    if request.method != "POST":
        return (json.dumps({"error": "Only POST allowed"}), 405, {"Content-Type": "application/json"})

    try:
        data = request.json
        ticker = data.get("ticker", "").upper()
        min_dte = int(data.get("min_dte", 30))
        max_dte = int(data.get("max_dte", 180))
        strategy = data.get("strategy", "bullish")

        if not ticker or strategy not in ["bullish", "bearish"]:
            return (json.dumps({"error": "Invalid input"}), 400, {"Content-Type": "application/json"})

        stock = yf.Ticker(ticker)
        time.sleep(0.5)
        hist = stock.history(period="1y")
        time.sleep(0.5)
        options_dates = stock.options
        time.sleep(0.5)

        # For demo: just use the first expiry
        expiry = options_dates[0]
        chain = stock.option_chain(expiry)
        time.sleep(0.5)

        if strategy == "bullish":
            options = chain.calls
            options = options.sort_values("impliedVolatility", ascending=False).head(5)
        else:
            options = chain.puts
            options = options.sort_values("impliedVolatility", ascending=False).head(5)

        # Prepare top 5 table
        top_5 = []
        for _, row in options.iterrows():
            top_5.append([
                row["contractSymbol"],
                str(row["strike"]),
                str(row["lastPrice"]),
                str(row["impliedVolatility"]),
                str(row["openInterest"]),
                str(row["volume"]),
                expiry
            ])

        # Prepare chart data (dummy P&L curve)
        chart_data = []
        strikes = [float(row[1]) for row in top_5]
        for price in np.linspace(min(strikes)*0.9, max(strikes)*1.1, 20):
            profit = (price - strikes[0]) * 10 if strategy == "bullish" else (strikes[0] - price) * 10
            chart_data.append({"price": round(price,2), "profit": round(profit,2)})

        summary = f"Top {strategy.capitalize()} trade for {ticker}: {top_5[0][0]} (Strike {top_5[0][1]}, IV {top_5[0][3]})"

        return (json.dumps({
            "summary": summary,
            "top_5": top_5,
            "chartData": chart_data
        }), 200, {"Content-Type": "application/json"})

    except Exception as e:
        return (json.dumps({"error": str(e)}), 500, {"Content-Type": "application/json"}) 